gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/verification/StandardVerificationStrategy.java

```diff
@@ -63,15 +64,17 @@ public void apply(final Traversal.Admin<?, ?> traversal) {
                 throw new VerificationException("The parent of a reducing barrier can not be repeat()-step: " + step, traversal);
         }
 
-        // The ProfileSideEffectStep must be the last step, 2nd last step when accompanied by the cap step,
-        // or 3rd to last when the traversal ends with a RequirementsStep.
+        // The ProfileSideEffectStep must be one of the following
+        // (1) the last step
+        // (2) 2nd last step when accompanied by the cap step or none step (i.e. iterate() to nothing)
+        // (3) 3rd to last when the traversal ends with a RequirementsStep.
         final Step<?, ?> endStep = traversal.asAdmin().getEndStep();
         if (TraversalHelper.hasStepOfClass(ProfileSideEffectStep.class, traversal) &&
                 !(endStep instanceof ProfileSideEffectStep ||
                         (endStep instanceof SideEffectCapStep && endStep.getPreviousStep() instanceof ProfileSideEffectStep) ||
-                        (endStep instanceof RequirementsStep && (
-                                endStep.getPreviousStep() instanceof SideEffectCapStep ||
-                                        endStep.getPreviousStep() instanceof ProfileSideEffectStep)))) {
+                        (endStep instanceof NoneStep && endStep.getPreviousStep() instanceof SideEffectCapStep && endStep.getPreviousStep().getPreviousStep() instanceof ProfileSideEffectStep) ||
```

Comment:
I think we can save some `instanceof` checks if we introduce a method similar to `TraversalHelper.hasStepOfClass(ProfileSideEffectStep.class, traversal)` but instead of returning a boolean, it returns the step instance. Then, from the `ProfileSideEffectStep` instance itself, it's easier to do the checks.

Reply:
yeah - @dkuppitz is thinking about this kind of stuff. i had ideas too, but figured it was something for a different issue. Generally speaking, I think that if we want to do something about the complexity of strategy writing, we really need to carefully analyze the strategies that are out there and find good general functions to add that can be best used across all of them rather than looking at just a bad problem in one and build a function that only fits that case.